app/src/main/java/org/mozilla/fenix/historymetadata/HistoryMetadataFeature.kt

```diff
@@ -4,77 +4,124 @@
-                                totalViewTime = acc.totalViewTime + elem.totalViewTime,
-                                updatedAt = max(acc.updatedAt, elem.updatedAt)
-                            )
+            val recentHistoryGroups = async {
+                historyMetadataStorage.getHistoryMetadataSince(Long.MIN_VALUE)
+                    .filter { it.totalViewTime > 0 && it.key.searchTerm != null }
+                    .groupBy { it.key.searchTerm!! }
+                    .mapValues { group ->
+                        // Within a group, we dedupe entries based on their url so we don't display
+                        // a page multiple times in the same group, and we sum up the total view time
+                        // of deduped entries while making sure to keep the latest updatedAt value.
+                        val metadataInGroup = group.value
+                        val metadataUrlGroups = metadataInGroup.groupBy { metadata -> metadata.key.url }
+                        metadataUrlGroups.map { metadata ->
+                            metadata.value.reduce { acc, elem ->
+                                acc.copy(
+                                    totalViewTime = acc.totalViewTime + elem.totalViewTime,
+                                    updatedAt = max(acc.updatedAt, elem.updatedAt)
+                                )
+                            }
                         }
                     }
-                }
-                .map { (title, data) ->
-                    HistoryMetadataGroup(
-                        title = title,
-                        historyMetadata = data
-                    )
-                }
-                .sortedByDescending { it.lastUpdated() }
-                .take(maxResults)
+                    .map { (title, data) ->
+                        RecentHistoryGroup(
+                            title = title,
+                            historyMetadata = data
+                        )
+                    }
+                    .sortedByDescending { it.lastUpdated() }
+                    .take(MAX_RESULTS_OF_EACH_TYPE)
+            }
 
-            homeStore.dispatch(HomeFragmentAction.HistoryMetadataChange(historyMetadata))
+            updateState(historyHighlights.await(), recentHistoryGroups.await())
         }
     }
 
+    @VisibleForTesting
+    internal fun updateState(
+        historyHighlights: List<RecentHistoryHighlight>,
+        historyGroups: List<RecentHistoryGroup>
+    ) {
+        homeStore.dispatch(
+            HomeFragmentAction.RecentHistoryChange(
+                getCombinedHistory(historyHighlights, historyGroups)
+            )
+        )
+    }
+
+    @VisibleForTesting
+    @Suppress("MagicNumber")
+    internal fun getCombinedHistory(
+        historyHighlights: List<RecentHistoryHighlight>,
+        historyGroups: List<RecentHistoryGroup>,
+    ): List<RecentlyVisitedItem> {
+        val recentGroups = historyGroups.take(3)
```

Comment:
Maybe add another const to for the 3 to remove the MagicNumber.

Reply:
Was thinking that this is not really part of the specs but can add it above but just a temporary approach.
Will add it as a top level const.

Reply:
Moved away from this logic in the latest version of the patch.